pkg/controller/bootstrap/bootstrapsigner.go

```diff
@@ -0,0 +1,291 @@
+	// If zero, a default max is used
+	MaxRetries int
+}
+
+// DefaultBootstrapSignerOptions returns a set of default options for creating a
+// BootstrapSigner
+func DefaultBootstrapSignerOptions() BootstrapSignerOptions {
+	return BootstrapSignerOptions{
+		ConfigMapNamespace: api.NamespacePublic,
+		ConfigMapName:      configMapClusterInfo,
+	}
+}
+
+// BootstrapSigner is a controller that signs a ConfigMap with a set of tokens.
+type BootstrapSigner struct {
+	stopChan chan struct{}
+
+	client       clientset.Interface
+	configMapKey string
+
+	configMaps cache.Store
+	secrets    cache.Store
+
+	// Since we join two objects, we'll watch both of them with controllers.
+	configMapsController *cache.Controller
+	secretsController    *cache.Controller
+
+	maxRetries int
+}
+
+// NewBootstrapSigner returns a new *ServiceAccountsController.
+func NewBootstrapSigner(cl clientset.Interface, options BootstrapSignerOptions) *BootstrapSigner {
+	maxRetries := options.MaxRetries
+	if maxRetries == 0 {
+		maxRetries = 10
+	}
+
+	e := &BootstrapSigner{
+		client:       cl,
+		configMapKey: fmt.Sprintf("%s/%s", options.ConfigMapNamespace, options.ConfigMapName),
+		maxRetries:   maxRetries,
+	}
+	if cl != nil && cl.Core().RESTClient().GetRateLimiter() != nil {
+		metrics.RegisterMetricAndTrackRateLimiterUsage("bootstrap_signer", cl.Core().RESTClient().GetRateLimiter())
+	}
+	configMapSelector := fields.SelectorFromSet(map[string]string{api.ObjectNameField: options.ConfigMapName})
+	e.configMaps, e.configMapsController = cache.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func(lo api.ListOptions) (runtime.Object, error) {
+				lo.FieldSelector = configMapSelector
+				return e.client.Core().ConfigMaps(options.ConfigMapNamespace).List(lo)
+			},
+			WatchFunc: func(lo api.ListOptions) (watch.Interface, error) {
+				lo.FieldSelector = configMapSelector
+				return e.client.Core().ConfigMaps(options.ConfigMapNamespace).Watch(lo)
+			},
+		},
+		&api.ConfigMap{},
+		options.ConfigMapResync,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    func(_ interface{}) { e.signConfigMap() },
+			UpdateFunc: func(_, _ interface{}) { e.signConfigMap() },
+		},
+	)
+
+	secretSelector := fields.SelectorFromSet(map[string]string{api.SecretTypeField: string(api.SecretTypeBootstrapToken)})
+	e.secrets, e.secretsController = cache.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func(lo api.ListOptions) (runtime.Object, error) {
+				lo.FieldSelector = secretSelector
+				return e.client.Core().Secrets(api.NamespaceSystem).List(lo)
+			},
+			WatchFunc: func(lo api.ListOptions) (watch.Interface, error) {
+				lo.FieldSelector = secretSelector
+				return e.client.Core().Secrets(api.NamespaceSystem).Watch(lo)
+			},
+		},
+		&api.Secret{},
+		options.SecretResync,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    func(_ interface{}) { e.signConfigMap() },
+			UpdateFunc: func(_, _ interface{}) { e.signConfigMap() },
```

Comment:
why not accept a config map from signConfigMap() ? Then you only check configmaps that chaned.